File: csharp/MkCert/KeyUtils.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Org.BouncyCastle.Crypto;
using Org.BouncyCastle.Crypto.Parameters;
using Org.BouncyCastle.OpenSsl;
using Org.BouncyCastle.Math;
using System.IO;

namespace MkCert
{
    public class KeyUtils
    {
        public static AsymmetricCipherKeyPair GenerateKeyPair()
        {
            var keyPairGenerator = new Org.BouncyCastle.Crypto.Generators.RsaKeyPairGenerator();
            keyPairGenerator.Init(new RsaKeyGenerationParameters(new BigInteger("65537"),new Org.BouncyCastle.Security.SecureRandom(), 2048, 80));
            return keyPairGenerator.GenerateKeyPair();
        }


        public static string RsaKeyToPem(RsaKeyParameters keyParams)
        {
            using (MemoryStream ms = new MemoryStream())
            {
                PemWriter writer = new PemWriter(new StreamWriter(ms));
                if(keyParams.IsPrivate)
                {
                    Pkcs8Generator generator = new Pkcs8Generator((RsaPrivateCrtKeyParameters)keyParams);
                    writer.WriteObject(generator.Generate());
                }
                else
                {
                    writer.WriteObject(keyParams);
                }
                writer.Writer.Flush();
                return System.Text.Encoding.ASCII.GetString(ms.ToArray());
            }
        }

        public static RsaKeyParameters RsaKeyFromPem(string keyPem)
        {
            using (MemoryStream ms = new MemoryStream(System.Text.Encoding.ASCII.GetBytes(keyPem)))
            {
                var pemReader = new PemReader(new StreamReader(ms));
                return pemReader.ReadObject() as RsaKeyParameters;
            }
        }
    }
}

File: java/src/main/java/KeyUtils.java


import org.bouncycastle.asn1.pkcs.EncryptedPrivateKeyInfo;
import org.bouncycastle.asn1.pkcs.PrivateKeyInfo;
import org.bouncycastle.asn1.x509.SubjectPublicKeyInfo;
import org.bouncycastle.crypto.AsymmetricCipherKeyPair;
import org.bouncycastle.crypto.params.RSAKeyGenerationParameters;
import org.bouncycastle.crypto.generators.RSAKeyPairGenerator;
import org.bouncycastle.crypto.params.AsymmetricKeyParameter;
import org.bouncycastle.crypto.params.RSAPrivateCrtKeyParameters;
import org.bouncycastle.crypto.params.RSAKeyParameters;
import org.bouncycastle.crypto.util.PrivateKeyFactory;
import org.bouncycastle.crypto.util.PrivateKeyInfoFactory;
import org.bouncycastle.crypto.util.PublicKeyFactory;
import org.bouncycastle.crypto.util.SubjectPublicKeyInfoFactory;
import org.bouncycastle.util.io.pem.PemReader;
import org.bouncycastle.util.io.pem.PemWriter;
import org.bouncycastle.util.io.pem.PemObject;

import java.io.IOException;
import java.io.StringReader;
import java.io.StringWriter;
import java.math.BigInteger;
import java.security.SecureRandom;

public class KeyUtils {

    public static AsymmetricCipherKeyPair generateKeyPair() {
        RSAKeyPairGenerator keyPairGenerator = new RSAKeyPairGenerator();
        keyPairGenerator.init(new RSAKeyGenerationParameters(new BigInteger("65537"),new SecureRandom(),2048,80));
        return keyPairGenerator.generateKeyPair();
    }

    public static  String RSAPrivateKeyToPem(RSAPrivateCrtKeyParameters privateKeyParam) throws IOException {
        StringWriter pemStrWriter = new StringWriter();
        PemWriter pemWriter = new PemWriter(pemStrWriter);
        PrivateKeyInfo privateKeyInfo = PrivateKeyInfoFactory.createPrivateKeyInfo(privateKeyParam);
        pemWriter.writeObject(new PemObject("PRIVATE KEY", privateKeyInfo.getEncoded()));
        pemWriter.flush();
        return pemStrWriter.toString();
    }

    public static String RSAPublicKeyToPem(RSAKeyParameters publicKeyParam) throws IOException {
        StringWriter pemStrWriter = new StringWriter();
        PemWriter pemWriter = new PemWriter(pemStrWriter);
        SubjectPublicKeyInfo subjectPublicKeyInfo = SubjectPublicKeyInfoFactory.createSubjectPublicKeyInfo(publicKeyParam);
        pemWriter.writeObject(new PemObject("PUBLIC KEY", subjectPublicKeyInfo.getEncoded()));
        pemWriter.flush();
        return pemStrWriter.toString();
    }

    public static  RSAPrivateCrtKeyParameters RSAPrivateKeyFromPem(String privateKeyPem) throws IOException {
        StringReader sr = new StringReader(privateKeyPem);
        PemReader pemReader = new PemReader(sr);
        return (RSAPrivateCrtKeyParameters) PrivateKeyFactory.createKey(pemReader.readPemObject().getContent());
    }

    public static  RSAKeyParameters RSAPublicKeyFromPem(String publicKeyPem) throws IOException {
        StringReader sr = new StringReader(publicKeyPem);
        PemReader pemReader = new PemReader(sr);
        return (RSAKeyParameters)PublicKeyFactory.createKey(pemReader.readPemObject().getContent());
    }
}